codec/src/main/java/io/netty5/handler/codec/MessageToByteEncoderForBuffer.java

```diff
@@ -0,0 +1,155 @@
+        this(outboundMessageType, true);
+    }
+
+    /**
+     * Create a new instance which will try to detect the types to match out of the type parameter of the class.
+     *
+     * @param preferDirect {@code true} if a direct {@link Buffer} should be tried to be used as target for
+     *                     the encoded messages. If {@code false} is used it will allocate a heap {@link Buffer}.
+     */
+    protected MessageToByteEncoderForBuffer(boolean preferDirect) {
+        matcher = TypeParameterMatcher.find(this, MessageToByteEncoderForBuffer.class, "I");
+        this.preferDirect = preferDirect;
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * @param outboundMessageType The type of messages to match.
+     * @param preferDirect        {@code true} if a direct {@link Buffer} should be tried to be used as target for
+     *                            the encoded messages. If {@code false} is used it will allocate a heap {@link Buffer}.
+     */
+    protected MessageToByteEncoderForBuffer(Class<? extends I> outboundMessageType, boolean preferDirect) {
+        matcher = TypeParameterMatcher.get(outboundMessageType);
+        this.preferDirect = preferDirect;
+    }
+
+    /**
+     * Returns {@code true} if the given message should be handled. If {@code false} it will be passed to the next
+     * {@link ChannelHandler} in the {@link ChannelPipeline}.
+     */
+    public boolean acceptOutboundMessage(Object msg) throws Exception {
+        return matcher.match(msg);
+    }
+
+    @Override
+    public Future<Void> write(ChannelHandlerContext ctx, Object msg) {
+        Buffer buf = null;
+        try {
+            if (acceptOutboundMessage(msg)) {
+                @SuppressWarnings("unchecked")
+                I cast = (I) msg;
+                buf = allocateBuffer(ctx, cast, preferDirect);
+                try {
+                    encode(ctx, cast, buf);
+                } finally {
+                    ReferenceCountUtil.release(cast);
+                }
+
+                if (buf.readableBytes() > 0) {
+                    Future<Void> f = ctx.write(buf);
+                    buf = null;
+                    return f;
+                }
+                return ctx.write(ctx.bufferAllocator().allocate(0));
+            }
+            return ctx.write(msg);
+        } catch (EncoderException e) {
+            return ctx.newFailedFuture(e);
+        } catch (Throwable e) {
+            return ctx.newFailedFuture(new EncoderException(e));
+        } finally {
+            if (buf != null) {
+                buf.close();
+            }
+        }
+    }
+
+    /**
+     * Allocate a {@link Buffer} which will be used as argument of {@link #encode(ChannelHandlerContext, I, Buffer)}.
+     * Sub-classes may override this method to return {@link Buffer} with a perfect matching {@code size}.
+     */
+    protected Buffer allocateBuffer(@SuppressWarnings("unused") ChannelHandlerContext ctx,
```

Comment:
Alternatively, if this method became abstract, there wouldn't be a need for the `preferDirect` parameter, since the choice of allocator would determine what kind of buffer was allocated. If sub-classes want to make that configurable, they can introduce their own `preferDirect` setting.

Reply:
Good idea!

Reply:
Agree let me do that